cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go

```diff
@@ -32,121 +37,159 @@ import (
 	rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1"
 )
 
+type updateKubeadmConfigMap func() error
+
 const (
 	// NodesKubeadmConfigClusterRoleName sets the name for the ClusterRole that allows
 	// the bootstrap tokens to access the kubeadm-config ConfigMap during the node bootstrap/discovery
 	// or during upgrade nodes
 	NodesKubeadmConfigClusterRoleName = "kubeadm:nodes-kubeadm-config"
 )
 
+var (
+	lockIdentity = string(uuid.NewUUID())
+)
+
 // ResetClusterStatusForNode removes the APIEndpoint of a given control-plane node
 // from the ClusterStatus and updates the kubeadm ConfigMap
-func ResetClusterStatusForNode(nodeName string, client clientset.Interface) error {
+func ResetClusterStatusForNode(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
 	fmt.Printf("[reset] Removing info for node %q from the ConfigMap %q in the %q Namespace\n",
-		nodeName, kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+		cfg.NodeRegistration.Name, kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+
+	return exclusivelyUpdateKubeadmConfigMap(client, func() error {
+		return removeCurrentEndpointFromClusterConfiguration(cfg, client)
+	})
+}
 
-	// Get the kubeadm ConfigMap
-	configMap, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(kubeadmconstants.KubeadmConfigConfigMap, metav1.GetOptions{})
+func exclusivelyUpdateKubeadmConfigMap(client clientset.Interface, updateConfigMap updateKubeadmConfigMap) error {
+	lock, err := resourcelock.New(resourcelock.ConfigMapsResourceLock, metav1.NamespaceSystem, kubeadmconstants.KubeadmConfigConfigMap, client.CoreV1(), nil, resourcelock.ResourceLockConfig{
+		Identity: lockIdentity,
+	})
 	if err != nil {
-		return errors.Wrap(err, "failed to get config map")
+		return errors.Wrap(err, "failed to create a configmap resource lock")
 	}
 
-	// Handle missing ClusterConfiguration in the ConfigMap. Should only happen if someone manually
-	// interacted with the ConfigMap.
-	clusterConfigurationYaml, ok := configMap.Data[kubeadmconstants.ClusterConfigurationConfigMapKey]
-	if !ok {
-		return errors.Errorf("cannot find key %q in ConfigMap %q in the %q Namespace",
-			kubeadmconstants.ClusterConfigurationConfigMapKey, kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+	ctx, cancel := context.WithCancel(context.TODO())
+	res := make(chan error, 1)
+
+	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
```

Comment:
Do you see the logs if you are waiting for locks?

Reply:
Yes, the logs are shown.